### PR TITLE
Phase 3: Conformance and drift hardening

### PR DESCRIPTION
**Pull request by jeremy**
## Summary
- Fix two no-op conformance assertion stubs (`responseMeta.totalCount`, `urlOrigin`) — both now validate or fail loudly
- Implement `scripts/drift-check-reverse` — verifies every JSON-capable route is modeled or excluded
- Expand `excluded-routes.json` from 3 to 809 entries with categorized reasons
- Update Makefile to replace drift-check-reverse TODO stub

## Test plan
- [x] `make conformance-mvp` passes (78/78 tests)
- [x] `make drift-check-reverse` passes (846 routes checked)
- [x] `make drift-check-full` passes

Stack: 3/5 (depends on #6)

---
## Summary by cubic
Strengthens conformance checks and adds a reverse drift guard so every JSON-capable route is modeled or explicitly excluded. Converts two assertion stubs into real validations and wires the check into the Makefile.

- **New Features**
  - Added `scripts/drift-check-reverse` to verify all JSON-capable routes are covered or excluded.
  - `make drift-check-reverse` now runs the script.
  - Expanded `spec/excluded-routes.json` from 3 to 809 entries with categorized reasons.

- **Bug Fixes**
  - `responseMeta.totalCount` now validates the `X-Total-Count` header and matches it to the expected integer.
  - `urlOrigin` parses `Link` headers with a robust `<...>` scanner, detects cross-origin next URLs for rejection tests, and fails on unsupported expected values.
  - Safer assertion handling: `toInt` rejects non-integer float64 values, adds type checks, and improves failure messages.
  - `scripts/drift-check-reverse` hardening: Bash 4+ check, anchored `^/jobs` filter, guard when the jq filter yields zero routes, and validate all JSON inputs (`route-snapshot.json`, `route-coverage.json`, `excluded-routes.json`) before length checks for clearer errors.

<sup>Written for commit f66dd83d10f7236032309d27b689e6316ce0bd6a. Summary will update on new commits.</sup>

### CODE REVIEW

File: conformance/runner/go/main.go [hunk 1]
```diff
@@ -432,22 +435,54 @@ func checkAssertion(testName string, a Assertion, s checkState) TestResult {
...
+				return fail(testName, "X-Total-Count header not present in response")
+			}
+			expected := int(a.Expected.(float64))
+			actual, err := strconv.Atoi(header)
+			if err != nil {
+				return fail(testName, "X-Total-Count header %q is not a valid integer", header)
+			}
+			if actual != expected {
+				return fail(testName, "Expected X-Total-Count=%d, got %d", expected, actual)
+			}
```
[the remark this answers is not in the document]
Comment (jeremy) [reply conformance/runner/go/main.go:485]: Fixed — added toInt() helper with type switch handling float64, int, json.Number, and string

File: conformance/runner/go/main.go [hunk 2]
```diff
@@ -432,22 +435,54 @@ func checkAssertion(testName string, a Assertion, s checkState) TestResult {
...
+				return fail(testName, "Failed to parse Link URL %q: %v", nextURL, err)
+			}
+			if linkParsed.IsAbs() && !strings.EqualFold(linkParsed.Host, serverURL.Host) {
+				// Cross-origin Link URL confirms the test scenario for rejection
+			} else if !linkParsed.IsAbs() {
+				return fail(testName, "Expected cross-origin Link URL for rejection test, but got relative URL: %s", nextURL)
+			} else if strings.EqualFold(linkParsed.Scheme, serverURL.Scheme) {
+				return fail(testName, "Expected cross-origin Link URL for rejection test, but %s has same origin as server", nextURL)
+			}
+		}
```
[the remark this answers is not in the document]
Comment (jeremy) [reply conformance/runner/go/main.go]: Fixed — safe type assertion via comma-ok; unknown expected values now fail loudly

File: conformance/runner/go/main.go [hunk 3]
```diff
@@ -464,6 +499,20 @@ func fail(testName, format string, args ...interface{}) TestResult {
...
+	for _, part := range strings.Split(linkHeader, ",") {
+		part = strings.TrimSpace(part)
+		if strings.Contains(part, `rel="next"`) {
+			start := strings.Index(part, "<")
+			end := strings.Index(part, ">")
+			if start >= 0 && end > start {
+				return part[start+1 : end]
+			}
+		}
+	}
```
[the remark this answers is not in the document]
Comment (jeremy) [reply conformance/runner/go/main.go:601]: Fixed — extractNextLinkURL now scans for <...> blocks instead of splitting on commas

File: scripts/drift-check-reverse [hunk 4]
```diff
@@ -0,0 +1,151 @@
...
+# "JSON-capable" means the route is part of the app's API surface — it can
+# serve JSON responses to API clients. Routes that are purely HTML UI, auth
+# flows, infrastructure, or admin are filtered out.
+#
+# Exit codes:
+#   0 = All JSON-capable routes accounted for
+#   1 = Unaccounted routes detected
+
+set -euo pipefail
+
```
[the remark this answers is not in the document]
Comment (jeremy) [reply scripts/drift-check-reverse:16]: Fixed — added Bash 4+ version check at script top

File: scripts/drift-check-reverse [hunk 5]
```diff
@@ -0,0 +1,151 @@
...
+  local method path
+  method=$(echo "$1" | tr '[:upper:]' '[:lower:]')
+  path=$(echo "$2" | sed -E 's/\{[^}]+\}/{_}/g')
+  path=$(echo "$path" | sed -E 's/\.json$//')
+  echo "$method $path"
+}
+
+# ---- Build lookup sets from coverage + excluded ----
+
+declare -A accounted_set
```
[the remark this answers is not in the document]
Comment (jeremy) [reply scripts/drift-check-reverse:61]: Fixed — same as above, version check covers both declare -A usages

File: conformance/runner/go/main.go [hunk 6]
```diff
@@ -432,22 +435,54 @@ func checkAssertion(testName string, a Assertion, s checkState) TestResult {
...
 		default:
 			return fail(testName, "Unknown responseMeta path: %s", a.Path)
 		}
 
 	case "urlOrigin":
-		// urlOrigin assertions verify pagination URL origin validation.
-		// The generated client doesn't handle pagination directly (that's hey.Client).
-		// For now, we verify the mock server sent only the expected number of requests.
-		// The requestCount assertion handles the validation indirectly.
+		expected := a.Expected.(string)
```
[the remark this answers is not in the document]
Comment (jeremy) [reply conformance/runner/go/main.go]: Fixed — urlOrigin now fails with unsupported value error for non-"rejected" expected values

File: scripts/drift-check-reverse [hunk 7]
```diff
@@ -0,0 +1,151 @@
...
+  # Exclude Queenbee (admin app), storefront, world (HEY World public routes)
+  select(.path | test("^/queenbee/|^/storefront/|^/world/") | not) |
+  # Exclude mail address changes (HTML confirmation flows)
+  select(.path | test("^/mail_address_change") | not) |
+  # Exclude ActiveStorage blob/representation routes
+  select(.path | test("/blobs/|/representations/|\\*filename") | not) |
+  # Exclude announcement management
+  select(.path | test("^/announcements") | not) |
+  # Exclude Solid Queue job UI routes
+  select(.path | test("/jobs|^/queues|^/applications/|^/recurring_tasks") | not) |
```
[the remark this answers is not in the document]
Comment (jeremy) [reply scripts/drift-check-reverse]: Fixed — anchored to `^/jobs` instead of unanchored `/jobs`

File: scripts/drift-check-reverse [hunk 8]
```diff
@@ -0,0 +1,151 @@
...
+  if [ -z "${accounted_set[$key]+x}" ]; then
+    echo "UNACCOUNTED: $method $path"
+    unaccounted=$((unaccounted + 1))
+  fi
+done < <(jq -r "$json_capable_filter" "$SNAPSHOT_FILE")
+
+echo ""
+echo "Checked $total JSON-capable routes."
+
+if [ "$unaccounted" -gt 0 ]; then
```
[the remark this answers is not in the document]
Comment (jeremy) [reply scripts/drift-check-reverse:157]: Fixed — sanity check added: exit 1 if filtered route count is zero

File: conformance/runner/go/main.go [hunk 9]
```diff
@@ -432,22 +435,54 @@ func checkAssertion(testName string, a Assertion, s checkState) TestResult {
...
+		if expected == "rejected" {
+			if s.sdkResp == nil {
```
[the remark this answers is not in the document]
Comment (jeremy) [reply conformance/runner/go/main.go]: Fixed — same as above, unsupported values fail loudly

File: scripts/drift-check-reverse [hunk 10]
```diff
@@ -0,0 +1,161 @@
...
+  if [ ! -f "$f" ]; then
+    echo "ERROR: $f not found."
+    exit 1
+  fi
+done
+
+if jq -e 'length == 0' "$SNAPSHOT_FILE" > /dev/null 2>&1; then
+  echo "ERROR: route-snapshot.json is empty. Run 'make drift-regen' to populate."
+  exit 1
+fi
```
[the remark this answers is not in the document]
Comment (jeremy) [reply scripts/drift-check-reverse]: Fixed — added explicit `jq empty` validation before the length check so malformed JSON fails immediately with a clear error.

File: conformance/runner/go/main.go [hunk 11]
```diff
@@ -464,6 +535,70 @@ func fail(testName, format string, args ...interface{}) TestResult {
 	}
 }
 
+// toInt safely converts an interface{} (typically from JSON) to int.
+func toInt(v interface{}) (int, error) {
+	switch n := v.(type) {
+	case float64:
+		return int(n), nil
```
[the remark this answers is not in the document]
Comment (jeremy) [reply conformance/runner/go/main.go:545]: Fixed — `toInt` now rejects non-integer float64 values: `if n != float64(int(n))` returns an error instead of silently truncating.

File: conformance/runner/go/main.go [hunk 12]
```diff
@@ -464,6 +535,70 @@ func fail(testName, format string, args ...interface{}) TestResult {
...
+		nextLink := strings.Index(rest, "<")
+		var params string
+		if nextLink >= 0 {
+			params = rest[:nextLink]
+		} else {
+			params = rest
+		}
+
+		if strings.Contains(params, `rel="next"`) {
+			return uri
```
[the remark this answers is not in the document]
Comment (jeremy) [reply conformance/runner/go/main.go:593]: Intentional scope — the conformance runner only tests against our own server responses which use standard `rel="next"` format. The Link header format is defined by the HEY API (matching Basecamp's existing pattern). Handling RFC 8288 edge cases like unquoted rel values or space-separated rel lists isn't needed here.

File: scripts/drift-check-reverse [hunk 13]
```diff
@@ -0,0 +1,165 @@
...
+done
+
+if ! jq empty "$SNAPSHOT_FILE" 2>/dev/null; then
+  echo "ERROR: route-snapshot.json is not valid JSON."
+  exit 1
+fi
+if jq -e 'length == 0' "$SNAPSHOT_FILE" > /dev/null; then
+  echo "ERROR: route-snapshot.json is empty. Run 'make drift-regen' to populate."
+  exit 1
+fi
```
[the remark this answers is not in the document]
Comment (jeremy) [reply scripts/drift-check-reverse]: Fixed — all three JSON files (route-snapshot, route-coverage, excluded-routes) are now validated with `jq empty` before processing.
